<br>Name: Shweta Dudhal</br>
<br>Course: BUDT704</br>
<br>Section: 0501</br>
<br>Date of Submission:22-Sept-2021</br>

# Card Game Analysis

A card game is played between two players. In total there are 25 rounds in one game.The rules of the games are as follows
<br>1-A player wins a game if his sum of cards is greater than his opponent's sum of cards</br>
<br>2-The game will stop if one of the player runs of the cards</br>
<br>3-If in a round , a player runs out of the cards the cards are given back to the player and player with the greater sum wins and the game stops</br>
<br>4-The game ends after 25 rounds</br>



A deck is created that consists of 36 cards and sum of the deck is also calculated.

In [1]:
##To create a Deck , so a game can be played among two opponents.
deck = list(range(2,11))*4
print(deck)
print(f'Length of deck:{len(deck)}')

[2, 3, 4, 5, 6, 7, 8, 9, 10, 2, 3, 4, 5, 6, 7, 8, 9, 10, 2, 3, 4, 5, 6, 7, 8, 9, 10, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Length of deck:36


In [2]:
## To calculate the total number of points available in the deck.
total=sum(deck)
print(f'Total points of the deck:{total}')

Total points of the deck:216


Imported random Library is used to shuffle the deck so that the cards can be distributed. The function shuffle is used to shuffle the elements. After that cards are distributed among the players

In [3]:
## Distribute equal number of cards to both the players
import random
def distribution (deck):
    p1=[]
    p2=[]
    random.shuffle(deck)
    for i in range(len(deck)):
        if i%2==0:
            p1.append(deck[i])
        else:
            p2.append(deck[i])
    #print(len(player_1))
    #print(len(player_2))
    return(p1,p2)


Below statement are used to print the no of cards and the sum that the player has.

In [4]:
## Invoke the function inorder to distribute the cards equally among both the players. 

cards_distributed=distribution(deck)
player_1_original=cards_distributed[0]
player_2_original=cards_distributed[1]
print(f'Overall cards that are distributed to both players:{cards_distributed}')
print(f'The cards distributed to player 1:{player_1_original}')
print(f'Player 1 has {len(player_1_original)} cards')
print(f'The cards distributed to player 2:{player_2_original}' )
print(f'Player 2 has {len(player_2_original)} cards')

total_player_1=sum(player_1_original)
total_player_2=sum(player_2_original)
print(f'Total Sum of cards of player 1 : {total_player_1}')
print(f'Total Sum of cards of player 2: {total_player_2}')

Overall cards that are distributed to both players:([5, 10, 3, 4, 9, 2, 9, 3, 10, 8, 4, 6, 5, 3, 3, 7, 9, 8], [2, 5, 5, 6, 2, 8, 7, 6, 10, 4, 7, 7, 2, 6, 10, 8, 4, 9])
The cards distributed to player 1:[5, 10, 3, 4, 9, 2, 9, 3, 10, 8, 4, 6, 5, 3, 3, 7, 9, 8]
Player 1 has 18 cards
The cards distributed to player 2:[2, 5, 5, 6, 2, 8, 7, 6, 10, 4, 7, 7, 2, 6, 10, 8, 4, 9]
Player 2 has 18 cards
Total Sum of cards of player 1 : 108
Total Sum of cards of player 2: 108


A function is created so that the games can be played according to the rule. Only one game is stimulated below, that consists of 25 rounds. It tells us the sum and length of the cards that players have. Also, the winner is declared.

In [5]:
#Function to play the game with proper conditions
def game(player_1,player_2):
    print(f'Player 1 cards:{player_1}')
    print(f'Player 2 cards:{player_2}')
    card_1_copy=[]
    card_2_copy=[]
    loop_breaker=0
    round=1
    max_round=25
    while (round<=max_round ):
        cards_1=[]
        cards_2=[]
        print(f'\nRound:{round}')
        # To show the first two cards 
        for i in range (0,len(cards_distributed)):
            cards_1.append(player_1[i])
            cards_2.append(player_2[i])  
            
        print(f'\nFirst two cards of player 1:{cards_1}')
        print(f'First two cards of player 2:{cards_2}')

        print(f'Sum of cards of player 1:{sum(cards_1)}')
        print(f'Sum of cards of player 2:{sum(cards_2)}')
      
        # To compare the sum of cards and decide the winner while adding cards to winner and deleting cards of the loser. 
        if sum(cards_1)>sum(cards_2):
            
            player_1=player_1[len(cards_1):] + player_1[:len(cards_1)]    ## for putting the cards showed by the player in the last
        
            for i in range(len(cards_2)):
                player_1.append(cards_2[i])
            player_2= player_2[len(cards_2):]          
           
            print("\nPlayer 1 Wins")
            print(f'New cards that player 1 has:{player_1}')
            print(f'New cards that player 2 has:{player_2}')
            print(f'No of cards of player 1 :{len(player_1)}')
            print(f'No of cards of player 2 :{len(player_2)}')
        

                    
        elif sum(cards_2)>sum(cards_1):
            
            player_2=player_2[len(cards_2):] + player_2[:len(cards_2)]   # for putting the cards showed by the player in the last
            for j in range(len(cards_1)):
                player_2.append(cards_1[j])
            player_1= player_1[len(cards_1):]     
            
            
            print("\nPlayer 2 Wins")
            print(f'New cards that player 1 has:{player_1}')
            print(f'New cards that player 2 has:{player_2}')
            print(f'No of cards of player 1 :{len(player_1)}')
            print(f'No of cards of player 2 :{len(player_2)}')
            
        
        else:
            player_1_copy=player_1.copy()
            player_2_copy=player_2.copy()
            cards_3=cards_1.copy()
            cards_4=cards_2.copy()
          
            while (True):
                if(sum(cards_3)==sum(cards_4)):  # To check if the sum of next two cards showed by the player are again equal
                        cards_3=[]
                        cards_4=[]
                        player_1_copy=player_1_copy[2:]
                        player_2_copy=player_2_copy[2:]
                        if((len(player_1_copy)==0) or (len(player_2_copy)==0)):
                            loop_breaker=1
                            break
                    
                        # To show next two cards as the sum of pervious cards were same
                        for j in range(0,len(cards_distributed)):
                            cards_1.append(player_1_copy[j])
                            cards_2.append(player_2_copy[j])
                            cards_3.append(player_1_copy[j])
                            cards_4.append(player_2_copy[j])
                        print(f'\nNext cards drawn by Player 1:{cards_3}')
                        print(f'Next cards drawn by Player 2:{cards_4}')
                        print(f'Sum of next cards drawn by Player 1:{sum(cards_3)}')
                        print(f'Sum of next cards drawn by Player 2:{sum(cards_4)}')
                      
                        # To check the sum of next two cards drawn and decide the winner
                        if sum(cards_3)>sum(cards_4):
                            print("\nPlayer 1 Wins")
                            for i in range(len(cards_1)):
                                player_1_copy.append(cards_1[i])
                            for i in range(len(cards_1)):
                                player_1_copy.append(cards_2[i])
                            player_2_copy=player_2_copy[len(cards_3):]
                            player_1_copy=player_1_copy[len(cards_3):]
                            player_1=player_1_copy
                            player_2=player_2_copy
                            print(f'\nNew cards that player 1 has:{player_1}')
                            print(f'New cards that player 2 has:{player_2}')
                            print(f'No of cards of player 1 :{len(player_1)}')
                            print(f'No of cards of player 2 :{len(player_2)}')
                 #           player_1_copy=[]
                  #          player_2_copy=[]
                           
                            break 

                        if sum(cards_4)>sum(cards_3):
                            print("Player 2 Wins")
                            for i in range(len(cards_1)):
                                player_2_copy.append(cards_2[i])
                                
                            for i in range(len(cards_2)):
                                player_2_copy.append(cards_1[i])
                            player_1_copy=(player_1_copy[len(cards_4):])
                            player_2_copy=(player_2_copy[len(cards_4):])
                            player_1=player_1_copy
                            player_2=player_2_copy
                            print(f'New cards that player 1 has:{player_1}')
                            print(f'New cards that player 2 has:{player_2}')
                            print(f'No of cards of player 1 :{len(player_1)}')
                            print(f'No of cards of player 2 :{len(player_2)}')
                    #        player_1_copy=[]
                     #       player_2_copy=[]
                            break
                                        
            
        
        round+=1 
        
       
    # To specify when the game should end.  
        if (len(player_1)==0 or len(player_2)==0) or loop_breaker==1:
            print('\nRan out of cards')
            break
        elif (round>max_round):
            print("\nRounds completed")
            
    # To determine the final scores of the winner and loser of the game. 
    if(sum(player_1)>sum(player_2)):
        print("\nWinner is Player 1")
        print("winner scores",sum(player_1))
        print("loser scores",sum(player_2))
        
    else:
        print("\nWinner is Player 2")
        print("winner scores",sum(player_2))
        print("loser scores",sum(player_1))
        
        
        
    print("Game Ends")            
            



In [6]:
#invoke the function to stimulate the game
game(player_1_original,player_2_original)

Player 1 cards:[5, 10, 3, 4, 9, 2, 9, 3, 10, 8, 4, 6, 5, 3, 3, 7, 9, 8]
Player 2 cards:[2, 5, 5, 6, 2, 8, 7, 6, 10, 4, 7, 7, 2, 6, 10, 8, 4, 9]

Round:1

First two cards of player 1:[5, 10]
First two cards of player 2:[2, 5]
Sum of cards of player 1:15
Sum of cards of player 2:7

Player 1 Wins
New cards that player 1 has:[3, 4, 9, 2, 9, 3, 10, 8, 4, 6, 5, 3, 3, 7, 9, 8, 5, 10, 2, 5]
New cards that player 2 has:[5, 6, 2, 8, 7, 6, 10, 4, 7, 7, 2, 6, 10, 8, 4, 9]
No of cards of player 1 :20
No of cards of player 2 :16

Round:2

First two cards of player 1:[3, 4]
First two cards of player 2:[5, 6]
Sum of cards of player 1:7
Sum of cards of player 2:11

Player 2 Wins
New cards that player 1 has:[9, 2, 9, 3, 10, 8, 4, 6, 5, 3, 3, 7, 9, 8, 5, 10, 2, 5]
New cards that player 2 has:[2, 8, 7, 6, 10, 4, 7, 7, 2, 6, 10, 8, 4, 9, 5, 6, 3, 4]
No of cards of player 1 :18
No of cards of player 2 :18

Round:3

First two cards of player 1:[9, 2]
First two cards of player 2:[2, 8]
Sum of cards of player 

# Data Analysis part -Stimulating function 50 times

To perform statistics analysis we need to import statistics library to use the functions available there. The function provides us the statistic summary of the game which will help us to gain insights and understand the game in a better way. Here the function created above is is stimulated 50 times i.e the game is played 50 times by the players.

In [7]:
#Remving all the print statements from the simulation fuction and adding logic to help with analysis 
def game_new(player_1,player_2):
    card_1_copy=[]
    card_2_copy=[]
    winning_value=0
    loop_breaker=0
    round=1
    max_round=25
    initial_points_p1=sum(player_1)
    initial_points_p2=sum(player_2)
    lowest_val=[]
    count=0
    while (round<=max_round ):
        cards_1=[]
        cards_2=[]
        # To show the first two cards 
        for i in range (0,len(cards_distributed)):
            cards_1.append(player_1[i])
            cards_2.append(player_2[i])  
      
        # To compare the sum of cards and decide the winner while adding cards to winner and deleting cards of the loser. 
        if sum(cards_1)>sum(cards_2):
            
            player_1=player_1[len(cards_1):] + player_1[:len(cards_1)]    ## for putting the cards showed by the player in the last
        
            for i in range(len(cards_2)):
                player_1.append(cards_2[i])
            player_2= player_2[len(cards_2):]          
        
        

                    
        elif sum(cards_2)>sum(cards_1):
            
            player_2=player_2[len(cards_2):] + player_2[:len(cards_2)]   # for putting the cards showed by the player in the last
            for j in range(len(cards_1)):
                player_2.append(cards_1[j])
            player_1= player_1[len(cards_1):]     
            
            
            
        
        else:
            player_1_copy=player_1.copy()
            player_2_copy=player_2.copy()
            cards_3=cards_1.copy()
            cards_4=cards_2.copy()
          
            while (True):
                if(sum(cards_3)==sum(cards_4)):  # To check if the sum of next two cards showed by the player are again equal
                        cards_3=[]
                        cards_4=[]
                        player_1_copy=player_1_copy[2:]
                        player_2_copy=player_2_copy[2:]
                        if((len(player_1_copy)==0) or (len(player_2_copy)==0)):
                            loop_breaker=1
                            break
                    
                        # To show next two cards as the sum of pervious cards were same
                        for j in range(0,len(cards_distributed)):
                            cards_1.append(player_1_copy[j])
                            cards_2.append(player_2_copy[j])
                            cards_3.append(player_1_copy[j])
                            cards_4.append(player_2_copy[j])
                      
                        # To check the sum of next two cards drawn and decide the winner
                        if sum(cards_3)>sum(cards_4):
                            for i in range(len(cards_1)):
                                player_1_copy.append(cards_1[i])
                            for i in range(len(cards_1)):
                                player_1_copy.append(cards_2[i])
                            player_2_copy=player_2_copy[len(cards_3):]
                            player_1_copy=player_1_copy[len(cards_3):]
                            player_1=player_1_copy
                            player_2=player_2_copy
                           
                            break 

                        if sum(cards_4)>sum(cards_3):
                            for i in range(len(cards_1)):
                                player_2_copy.append(cards_2[i])
                                
                            for i in range(len(cards_2)):
                                player_2_copy.append(cards_1[i])
                            player_1_copy=(player_1_copy[len(cards_4):])
                            player_2_copy=(player_2_copy[len(cards_4):])
                            player_1=player_1_copy
                            player_2=player_2_copy
                            break
                                        
            
        
        round+=1 
        
       
    # To specify when the game should end.  
        if (len(player_1)==0 or len(player_2)==0) or loop_breaker==1:
            break
        elif (round>max_round):
            break

            
    # To determine the final scores of the winner and loser of the game. 
    if(sum(player_1)>sum(player_2)):
        winning_value=sum(player_1)
        lowest_val=initial_points_p1
        if initial_points_p1>initial_points_p2:
            count=1
        
    else:
        winning_value=sum(player_2)
        lowest_val=initial_points_p2
        if initial_points_p2>initial_points_p1:
            count=1
            
    return((winning_value,lowest_val,count))

               
            



In [8]:
def analysis(deck):
    game_val=1
    winning_set=[]
    losing_set=[]
    lowest_val_set=[]
    count_set=[]
    while(game_val<=50):
        cards_distributed=distribution(deck)
        player_1=cards_distributed[0]
        player_2=cards_distributed[1]
        game_new_returns=game_new(player_1,player_2)
        winning_value=game_new_returns[0]
        lowest_val=game_new_returns[1]
        count=game_new_returns[2]
        winning_set.append(winning_value)
        losing_set.append(sum(deck)-winning_value)
        lowest_val_set.append(lowest_val)
        count_set.append(count)
        game_val+=1
    return((winning_set,losing_set,lowest_val_set,count_set))
    

In [9]:
analysis_values=analysis(deck)

In [10]:
print(analysis_values)

([212, 216, 141, 205, 120, 156, 125, 216, 115, 139, 160, 122, 144, 155, 168, 216, 128, 199, 216, 216, 121, 122, 180, 147, 125, 187, 131, 110, 190, 184, 216, 122, 127, 197, 216, 216, 216, 216, 115, 175, 207, 158, 216, 146, 140, 216, 216, 143, 216, 216], [4, 0, 75, 11, 96, 60, 91, 0, 101, 77, 56, 94, 72, 61, 48, 0, 88, 17, 0, 0, 95, 94, 36, 69, 91, 29, 85, 106, 26, 32, 0, 94, 89, 19, 0, 0, 0, 0, 101, 41, 9, 58, 0, 70, 76, 0, 0, 73, 0, 0], [122, 114, 110, 113, 103, 111, 110, 118, 114, 105, 110, 99, 115, 95, 115, 130, 105, 110, 122, 119, 98, 114, 107, 108, 97, 109, 116, 102, 112, 102, 124, 102, 103, 120, 115, 112, 112, 100, 102, 98, 111, 103, 116, 101, 103, 121, 113, 125, 128, 116], [1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1])


In [11]:
winning_set=analysis_values[0]
losing_set=analysis_values[1]
lowest_val_set=analysis_values[2]
count_set=analysis_values[3]

In [12]:
import statistics
print("Winning Set:",winning_set)
print()
print("Minimum of winning set:",min(winning_set))
print()
print("Maximum of winning set:",max(winning_set))
print()
print("Mean of winning set:",statistics.mean(winning_set))
print()
print("Losing Set:",losing_set)
print()
print("Minimum of losing set:",min(losing_set))
print()
print("Maximum of losing set:",max(losing_set))
print()
print("Mean of losing set:",statistics.mean(losing_set))


Winning Set: [212, 216, 141, 205, 120, 156, 125, 216, 115, 139, 160, 122, 144, 155, 168, 216, 128, 199, 216, 216, 121, 122, 180, 147, 125, 187, 131, 110, 190, 184, 216, 122, 127, 197, 216, 216, 216, 216, 115, 175, 207, 158, 216, 146, 140, 216, 216, 143, 216, 216]

Minimum of winning set: 110

Maximum of winning set: 216

Mean of winning set: 171.12

Losing Set: [4, 0, 75, 11, 96, 60, 91, 0, 101, 77, 56, 94, 72, 61, 48, 0, 88, 17, 0, 0, 95, 94, 36, 69, 91, 29, 85, 106, 26, 32, 0, 94, 89, 19, 0, 0, 0, 0, 101, 41, 9, 58, 0, 70, 76, 0, 0, 73, 0, 0]

Minimum of losing set: 0

Maximum of losing set: 106

Mean of losing set: 44.88


In [13]:
print("Lowest initial value for game win:",min(lowest_val_set))
print()
print(f"% times a game was won by player with lower initial value:{(sum(count_set)/50)*100}%")

Lowest initial value for game win: 95

% times a game was won by player with lower initial value:62.0%


### Inferences:

1.The mean value of the winning set is more towards its maximum value while the mean value of the losing et is more towards its minimum value showing that most of the people who win a match, win with a high score and those who lose, with a low score. We also see that the minimum value of winning set and the maximum value losing set is in the neighbourhood of 108. Thus a person cannot lose a game if his score is more than 108 and vice versa.

2.There is a 70-75% chances that a person starting with lower total card points ends up winning the game.This means that it is not necessary that a person who starts with a lower point value set always loses the match.But, this is also biased towards the person who starts the game with higher total card points even though the deck is evenly distributed at the start.

<br>"I pledge on my honor that I have not given nor received any unauthorized
assistance on this assignment."
--Shweta Dudhal</br>